### PR TITLE
Adding conditional sampling capability

### PR DESCRIPTION
**Pull request by zachmprince**
Refs #19350

### CODE REVIEW

**Comment by somu15** [comment]
@lynnmunday Could you please also take a look at this? @zachmprince implemented underlying base classes to support active learning in MOOSE STM. Thank you!

File: modules/stochastic_tools/include/reporters/ActiveLearningReporterBase.h [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+void
+ActiveLearningReporterTempl<T>::execute()
+{
+  // Dummy value in case _data has not been declared yet
+  T dummy;
+  // Loop over samples to determine if sample is needed. Replace value in _data
+  // (typically only if a sample is not needed). We insert a dummy value in case
+  // _data has not been declared.
+  for (const auto & i : make_range(_sampler.getNumberOfLocalRows()))
+    _need_sample[i] = needSample(_sampler.getNextLocalRow(),
```
Comment (somu15) [inline modules/stochastic_tools/include/reporters/ActiveLearningReporterBase.h:116]: I get it now. The reason you have two dof_id_type in need sample is two support distributed sampling, correct?

File: modules/stochastic_tools/src/reporters/ConditionalSampleReporter.C [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  if (!_disable_fpoptimizer)
+    _func_F->Optimize();
+  if (_enable_jit)
+    _func_F->JITCompile();
+  _func_params.resize(_sampler_vars.size() + (_use_time ? 1 : 0));
+}
+
+template <typename T>
+bool
+ConditionalSampleReporterTempl<T>::needSample(const std::vector<Real> & row,
```
Comment (somu15) [inline modules/stochastic_tools/src/reporters/ConditionalSampleReporter.C:83]: If we have to extend this to using a surrogate, instead of requesting a function from the user, we request a surrogate and compute the decision function in the `need_sample` itself.  Or else, we can have the surrogate itself in this active learning child class, as we discussed.

File: modules/stochastic_tools/src/reporters/ConditionalSampleReporter.C [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+                                              dof_id_type,
+                                              dof_id_type,
+                                              T & val)
+{
+  // Input sampler values
+  for (const auto & p : index_range(_sampler_vars))
```
Comment (somu15) [inline modules/stochastic_tools/src/reporters/ConditionalSampleReporter.C:89]: Another question is, can this type of child class access the subApp output if it is called? Accessing the subApp output will permit re-training the surrogate before making the prediction for the next sample.

File: modules/stochastic_tools/test/tests/multiapps/conditional_run/conditional_main.i [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+[]
+
+[Reporters]
+  [conditional]
+    type = ConditionalSampleReporter
+    sampler = cart
+    default_value = 1
+    function = 'val >= t'
+    sampler_vars = 'val'
+    sampler_var_indices = '0'
```
Comment (somu15) [inline modules/stochastic_tools/test/tests/multiapps/conditional_run/conditional_main.i:51]: Why is this required?

File: modules/stochastic_tools/test/tests/multiapps/conditional_run/conditional_main.i [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    parallel_type = ROOT
```
Comment (somu15) [inline modules/stochastic_tools/test/tests/multiapps/conditional_run/conditional_main.i:52]: So, the decision as to whether or not to call the subApp is only made on the root proc?

**Review by somu15 — COMMENTED** [review]
Thanks @zachmprince! I made some initial comments. I can take another look at it. Lynn would be a good person to review this as he is very good at multiapps.

File: modules/stochastic_tools/include/reporters/ActiveLearningReporterBase.h [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+template <typename T>
+class ActiveLearningReporterTempl;
+
+/**
+ * This is a base class for performing active learning routines, meant to be used
+ * in conjunction with Sampler multiapps and SamplerReporterTransfer. The purpose
+ * is to determine if a sample needs to be evaluated by a multiphysics solve and
+ * possibly replace quantities of interest with values computed by other means.
+ */
+template <typename T>
```
Comment (lynnmunday) [inline modules/stochastic_tools/include/reporters/ActiveLearningReporterBase.h:23]: Is the class templated only because _default_value is templated?   Is default_value ever anything other than a scalar?

File: modules/stochastic_tools/doc/content/source/reporters/ConditionalSampleReporter.md [hunk 7]
```diff
@@ -0,0 +1,19 @@
+# ConditionalSampleReporter
+
+!syntax description /Reporters/ConditionalSampleReporter
+
+## Description
+
+This object is a type of [StochasticReporter.md], so it can be used in to transfer reporters into using [SamplerReporterTransfer.md]. This object is mainly meant for demonstration for eventual active learning algorithms, but could prove useful. Basically, it enables a inputted function to determine if a multiapp solve is "possible" based on sampler values. The parsed function inputted with the parameter [!param](/Reporters/ConditionalSampleReporter/function) should evaluate to 0 or `false` if multiapp run is not possible/necessary. The variables within the function should be associated with a sampler column, which is defined by the pair of parameters [!param](/Reporters/ConditionalSampleReporter/sampler_vars) and [!param](/Reporters/ConditionalSampleReporter/sampler_var_indices). If the inputeed function evaluates to 0 or `false`, the quantity being transferred to this object using [SamplerReporterTransfer.md] with be replaced with [!param](/Reporters/ConditionalSampleReporter/default_value).
```
[the remark this answers is not in the document]
Comment (zachmprince) [reply modules/stochastic_tools/doc/content/source/reporters/ConditionalSampleReporter.md:7]: It is tie the column of the sampler matrix to a variable in the function. This is need in the case where not all the columns are used to evaluate the function.

File: modules/stochastic_tools/include/reporters/ActiveLearningReporterBase.h (hunk 6 above)
Comment (zachmprince) [reply modules/stochastic_tools/include/reporters/ActiveLearningReporterBase.h:23]: No, it is templated based on the type of reporter being transferred. We may want in the future to have a QoI be a vector, in such a case you will want to derive from the `ActiveLearningReporterTempl<std::vector<Real>>` version of this class.

File: modules/stochastic_tools/include/reporters/ActiveLearningReporterBase.h (hunk 1 above)
Comment (zachmprince) [reply modules/stochastic_tools/include/reporters/ActiveLearningReporterBase.h:116]: Correct.

File: modules/stochastic_tools/src/multiapps/SamplerFullSolveMultiApp.C [hunk 8]
```diff
@@ -64,6 +70,10 @@ SamplerFullSolveMultiApp::SamplerFullSolveMultiApp(const InputParameters & param
        _sampler.getRankConfig(_mode == StochasticTools::MultiAppMode::BATCH_RESET ||
                               _mode == StochasticTools::MultiAppMode::BATCH_RESTORE));
   _number_of_sampler_rows = _sampler.getNumberOfRows();
+
```
[the remark this answers is not in the document]
Comment (zachmprince) [reply modules/stochastic_tools/src/multiapps/SamplerFullSolveMultiApp.C:73]: I don't quite remember, but I believe you are correct. Another reason is that in normal mode, a sub-app is created for every sample, which might be a burden when most of them aren't even being run.

File: modules/stochastic_tools/src/multiapps/SamplerFullSolveMultiApp.C [hunk 9]
```diff
@@ -109,6 +122,15 @@ SamplerFullSolveMultiApp::solveStepBatch(Real dt, Real target_time, bool auto_ad
 {
   TIME_SECTION("solveStepBatch", 3, "Solving Step Batch For SamplerFullSolveMultiApp");
 
+  if (_should_run && _should_run->size() < _sampler.getNumberOfLocalRows())
+    paramError("should_run_reporter",
```
[the remark this answers is not in the document]
Comment (zachmprince) [reply modules/stochastic_tools/src/multiapps/SamplerFullSolveMultiApp.C:126]: You can technically put in any reporter value for this parameter, so just wanted to make it fool-proof. When using the ActiveLearning reporter, this should always be true though.

File: modules/stochastic_tools/src/reporters/ConditionalSampleReporter.C (hunk 2 above)
Comment (zachmprince) [reply modules/stochastic_tools/src/reporters/ConditionalSampleReporter.C:83]: I'm guessing you figured out the answer to this question already.

File: modules/stochastic_tools/src/reporters/ConditionalSampleReporter.C (hunk 3 above)
Comment (zachmprince) [reply modules/stochastic_tools/src/reporters/ConditionalSampleReporter.C:89]: I'm guessing you figured out the answer to this question already.

File: modules/stochastic_tools/test/tests/multiapps/conditional_run/conditional_main.i (hunk 4 above)
Comment (zachmprince) [reply modules/stochastic_tools/test/tests/multiapps/conditional_run/conditional_main.i:51]: It is to say the first column of the sampler matrix should be substituted in for `val` in the function.

File: modules/stochastic_tools/test/tests/multiapps/conditional_run/conditional_main.i (hunk 5 above)
Comment (zachmprince) [reply modules/stochastic_tools/test/tests/multiapps/conditional_run/conditional_main.i:52]: No, this is just for outputting. Makes it so the same json file gets created no matter how many processes are used.

File: modules/stochastic_tools/include/reporters/ActiveLearningReporterBase.h [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+                                                               const ReporterName & from_reporter,
+                                                               std::string prefix)
+{
+  // Only one value is allowed to be declared
+  if (_data)
+    this->mooseError(type(), " can only declare a single reporter value.");
+  // Make sure the inputted sampler is the same one in the parameters
+  else if (sampler.name() != _sampler.name())
+    this->paramError("sampler",
+                     "Inputted sampler, ",
```
Comment (grmnptr) [inline modules/stochastic_tools/include/reporters/ActiveLearningReporterBase.h:135]:
```suggestion
                     "Input sampler, ",
```

Comment (grmnptr) [reply modules/stochastic_tools/include/reporters/ActiveLearningReporterBase.h:135]: Feel free to discard this one!

File: modules/stochastic_tools/include/reporters/ConditionalSampleReporter.h [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+// Forward declaration
+template <typename T>
+class ConditionalSampleReporterTempl;
+
+// Typedef for object registration
+typedef ConditionalSampleReporterTempl<Real> ConditionalSampleReporter;
+
+/**
+ * This object is mainly meant for demonstration for eventual active learning
+ * algrorithms, but could prove useful. Basically, it enables a inputted function
```
Comment (grmnptr) [inline modules/stochastic_tools/include/reporters/ConditionalSampleReporter.h:24]:
```suggestion
 * algorithms, but could prove useful. Basically, it enables a inputted function
```

**Review by grmnptr — CHANGES_REQUESTED** [review]
Just a few typos. Nice job!

File: modules/stochastic_tools/doc/content/source/reporters/ConditionalSampleReporter.md [hunk 12]
```diff
@@ -0,0 +1,19 @@
+# ConditionalSampleReporter
+
+!syntax description /Reporters/ConditionalSampleReporter
+
+## Description
+
+This object is a type of [StochasticReporter.md], so it can be used to transfer reporter values from sub-apps into it using [SamplerReporterTransfer.md]. This object is mainly meant for demonstration for eventual active learning algorithms, but could prove useful. Basically, it enables a inputted function to determine if a multiapp solve is "possible" based on sampler values. The parsed function inputted with the parameter [!param](/Reporters/ConditionalSampleReporter/function) should evaluate to 0 or `false` if multiapp run is not possible/necessary. The variables within the function should be associated with a sampler column, which is defined by the pair of parameters [!param](/Reporters/ConditionalSampleReporter/sampler_vars) and [!param](/Reporters/ConditionalSampleReporter/sampler_var_indices). If the inputeed function evaluates to 0 or `false`, the quantity being transferred to this object using [SamplerReporterTransfer.md] with be replaced with [!param](/Reporters/ConditionalSampleReporter/default_value).
```
[the remark this answers is not in the document]
Comment (zachmprince) [reply modules/stochastic_tools/doc/content/source/reporters/ConditionalSampleReporter.md]: Two reasons. The first being the reason you mentioned. The second being the ability to determine which samples were skipped. If the element was deleted, it would be impossible to determine which sample row exactly caused a skip. Having a default value allows users to filter it when postprocessing.

[the remark this answers is not in the document]
Comment (zachmprince) [reply modules/stochastic_tools/doc/content/source/reporters/ConditionalSampleReporter.md]: Two reasons. The first being the reason you mentioned. The second being the ability to determine which samples were skipped. If the element was deleted, it would be impossible to determine which sample row exactly caused a skip. Having a default value allows users to filter it when postprocessing.

**Review by grmnptr — CHANGES_REQUESTED** [review]
Typos in the documentation, then it is ready to go.

**Review by grmnptr — APPROVED** [review]
